### PR TITLE
[SYCL] Fix resource leak related to SYCL_FALLBACK_ASSERT

### PR DESCRIPTION
**Pull request by aelovikov-intel**
https://github.com/intel/llvm/pull/6837 enabled asynchronous buffer destruction for buffers constructed without host data. However, initial fallback assert implementation in
https://github.com/intel/llvm/pull/3767 predates it and as such had to place the buffer inside `queue_impl` to avoid unintended synchronization point. I don't know if there was the same crash observed on the end-to-end test added as part of this PR prior to
https://github.com/intel/llvm/pull/3767, but it doesn't even matter because the "new" implementation is both simpler and doesn't result in a crash.

I suspect that without it (with the buffer for fallback assert implementation being a data member of `sycl::queue_impl`) we had a cyclic dependency somewhere leading to resource leak and ultimately to the assert in `DeviceGlobalUSMMem::~DeviceGlobalUSMMem()`.

### CODE REVIEW

**Comment by maarquitos14** [comment]
The newly added test is failing. Should I wait for further changes before review?

**Comment by aelovikov-intel** [comment]
My bet is those are yet another issue. AMD/HIP is totally different because it has `aspect::ext_oneapi_native_assert` and different code path is executed. Memory leaks detection on Windows is unreliable and I should disable it there.

The only remaining failure that could possibly be related is the leak on Lin/gen12 configuration, but even that is improvement over previous state where we'd fail with an assert because nothing was released. Here, at least, it's just one queue and one event.

**Comment by aelovikov-intel** [comment]
@maarquitos14 , I've removed the leak tracking from the test as those issues are distinct from what is being fixed here.

@KseniyaTikhomirova , @sergey-semenov I know you are busy but I think you're the most experienced in this area, would appreciate if you can take a quick look too.

**Review by sergey-semenov — APPROVED** [review]
LGTM, relying on the deferred release is much better.

**Review by maarquitos14 — APPROVED** [review]
LGTM.

File: sycl/include/sycl/queue.hpp [hunk 1]
```diff
@@ -3017,9 +3019,7 @@ class AssertInfoCopier;
  */
 event submitAssertCapture(queue &Self, event &Event, queue *SecondaryQueue,
                           const detail::code_location &CodeLoc) {
-  using AHBufT = buffer<detail::AssertHappened, 1>;
-
-  AHBufT &Buffer = Self.getAssertHappenedBuffer();
```
[the remark this answers is not in the document]
Comment (aelovikov-intel) [reply sycl/include/sycl/queue.hpp:3022]: Yes, exactly.

**Comment by aelovikov-intel** [comment]
Post commit failure (Arc GPU):
```
Failed Tests (1):
  SYCL :: Plugin/sycl-partition-info.cpp
```
```
FAIL: SYCL :: Plugin/sycl-partition-info.cpp (1486 of 1857)
******************** TEST 'SYCL :: Plugin/sycl-partition-info.cpp' FAILED ********************
Exit Code: -6

Command Output (stdout):
--
# RUN: at line 1
/__w/llvm/llvm/toolchain/bin//clang++   -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/Plugin/sycl-partition-info.cpp -o /__w/llvm/llvm/build-e2e/Plugin/Output/sycl-partition-info.cpp.tmp.out
# executed command: /__w/llvm/llvm/toolchain/bin//clang++ -fsycl -fsycl-targets=spir64 /__w/llvm/llvm/llvm/sycl/test-e2e/Plugin/sycl-partition-info.cpp -o /__w/llvm/llvm/build-e2e/Plugin/Output/sycl-partition-info.cpp.tmp.out
# note: command had no output on stdout or stderr
# RUN: at line 2
env ONEAPI_DEVICE_SELECTOR=level_zero:gpu  /__w/llvm/llvm/build-e2e/Plugin/Output/sycl-partition-info.cpp.tmp.out
# executed command: env ONEAPI_DEVICE_SELECTOR=level_zero:gpu /__w/llvm/llvm/build-e2e/Plugin/Output/sycl-partition-info.cpp.tmp.out
# .---command stdout------------
# | Abort was called at [25](https://github.com/intel/llvm/actions/runs/7731778500/job/21080773194#step:21:26)3 line in file:
# | ../../neo/level_zero/core/source/builtin/builtin_functions_lib_impl.cpp
# `-----------------------------
# .---command stderr------------
# | pure virtual method called
# | terminate called without an active exception
# `-----------------------------
# error: command failed with exit status: -6
```

**Pull request merged**